### PR TITLE
Debugger implementation

### PR DESCRIPTION
**Pull request by MartinWangIBM**
This commit contains both the frontend and the backend.

The frontend contains a debug service, a slideout pane in the side bar
(debug UI) and a deploy service.

The backend talks to debug adapters which implement VSCode Debug
Protocol. it can be used either as a module of Orionode or a standalone
server.

Change-Id: Ib6f17715f498301fa699bfad86a49ca6e667c7c9
Signed-off-by: Martin Wang <jiahaow@ca.ibm.com>

### CODE REVIEW

File: bundles/org.eclipse.orion.client.debug/web/orion/debug/nativeDeployService.js [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+    /**
+     * A debug service provides breakpoints, watches and highlighted lines management.
+     * Here management means loading and saving data from storage, apis of getters and setters, and correspoding events.
+     * Note that breakpoints from streaming (remote-only) files won't be presisted.
+     * 
+     * @class {orion.debug.NativeDeployService}
+     * @param {orion.serviceregistry.ServiceRegistry} serviceRegistry
+     * @param {orion.commandregistry.CommandRegistry} commandRegistry
+     */
+    var NativeDeployService = function(serviceRegistry, commandRegistry) {
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.debug/web/orion/debug/nativeDeployService.js:31]: I call the new deploy service for the debugger "Native deploy service". I don't think it's a good name, but other names I can imagine are even worse. Here are the reasons that I didn't select the following names:
Debug deploy service: You can also deploy a CF app with node-inspector and debug it.
Local deploy service: It's not local. The debugee runs on the same machine as Orion server runs.
Remote deploy service: All deployments are remote, aren't they? Also it confuse the "remote debug" checkbox in the configuration.

File: bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js [hunk 2]
```diff
@@ -216,13 +223,20 @@ define([
...
 						domNodeWrapperList = [];
 						this._commandRegistry.renderCommands(createNewItem.id, dropdownMenuItemSpan, this._project, this, "button", null, domNodeWrapperList); //$NON-NLS-0$
-						domNodeWrapperList[0].domNode.textContent = "+"; //$NON-NLS-0$
-						this._setNodeTooltip(domNodeWrapperList[0].domNode, messages["createNewTooltip"]); //$NON-NLS-0$
+						for (var i = 0; i < defaultDeployCommands.length; i++) {
+							domNodeWrapperList[i].domNode.textContent = defaultDeployCommands[i].name;
+							this._setNodeTooltip(domNodeWrapperList[i].domNode, defaultDeployCommands[i].tooltip);
+						}
+						if (defaultDeployCommands.length === 1) {
+							domNodeWrapperList[0].domNode.textContent = "+";
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js]: Here I hard coded that if there is only one deploy service (which is CF by default), the "new configuration" button will simply show "+" instead of its full name ("+CF" as I changed).
This keeps the consistency if debugger is disabled, but I think it's better to show "+CF" to tell the user that this is the configuration for a CF app.

File: bundles/org.eclipse.orion.client.debug/web/orion/debug/nls/root/messages.js [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * available under the terms of the Eclipse Public License v1.0 
+ * (http://www.eclipse.org/legal/epl-v10.html), and the Eclipse Distribution 
+ * License v1.0 (http://www.eclipse.org/org/documents/edl-v10.html). 
+ * 
+ ******************************************************************************/
+/*eslint-env browser, amd*/
+define({ //Default message bundle
+	"DebugWorkspace": "Debug Workspace",
+	"DebugHoverEvaluation": "Hover Evaluation",
+	"createNew": "+Native",
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.debug/web/orion/debug/nls/root/messages.js]: Any naming suggestion?

File: modules/orionode.debug.server/lib/debugFile.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
```
Comment (MartinWangIBM) [inline modules/orionode.debug.server/lib/debugFile.js:1]: This routing is not used but kept for debugging purpose.

File: modules/orionode.debug.server/adapters.json [hunk 5]
```diff
@@ -0,0 +1,57 @@
+{
```
Comment (MartinWangIBM) [inline modules/orionode.debug.server/adapters.json:1]: I put four adapters here as the "default" setup, because configuring an adapter is complicated. Adapters, especially the community ones, have different build command and many of them relies on VS Code specific functionalities such as executing extension commands. As a result, the user has to dig into the source code and find the right way to set up the adapter, which cost plenty of time.
However, I don't think we should be responsible for maintaining this configuration for community adapters.
Any suggestion?

File: bundles/org.eclipse.orion.client.ui/web/plugins/pageLinksPlugin.js [hunk 6]
```diff
@@ -43,15 +43,15 @@ define([
 			order: 10
 		});
 		
-		if (!util.isElectron) {
+		//if (!util.isElectron) {
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.ui/web/plugins/pageLinksPlugin.js]: Since we support debugger in Electron, shell should be enabled as well.

Do we want it? Should I put it in a separate patch?

File: bundles/org.eclipse.orion.client.ui/web/edit/setup.js [hunk 7]
```diff
@@ -1115,9 +1179,115 @@ objects.mixin(EditorViewer.prototype, {
 					editorViewer.updateDirtyIndicator();
 				});
 			}.bind(this));
+			this.editor.addEventListener("AnnotationModified", this.editorAnnotationModifiedListener = function(e) {
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.ui/web/edit/setup.js]: The current logic of adding a breakpoint is: double click on line numbers -> send annotation event -> convert annotation to breakpoint -> change breakpoints -> receive breakpoint event -> convert breakpoint to annotation -> add annotation.
It's weird and not good for an embedded editor.
Any suggestion?

**Comment by rcjsuen** [comment]
Where is the discussion about this feature supposed to take place? What is the associated bugzilla bug id?

File: bundles/org.eclipse.orion.client.debug/web/README.md [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        "templates": [
+            {
+                "SOME_KEY": "SOME_VALUE"
+            }
+        ]
+    }
+}
+~~~
+
+### Server Side Setup
```
Comment (rcjsuen) [inline bundles/org.eclipse.orion.client.debug/web/README.md:62]: I was looking for this setup information in the `modules/orionode.debug.server` folder. I imagine at some point the Node code will be split off the [`orion.server.node`](https://github.com/eclipse/orion.server.node) repository. I don't know if `orionode.debug.server` will go there too as the debug server can actually run both in the Java server and the Node server. With that in mind, could you also put installation instructions in the `modules/orionde.debug.server` folder?

File: modules/orionode.debug.server/package.json [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    "vscode-debugprotocol": "^1.16.0"
+  },
+  "devDependencies": {},
+  "engines": {
+    "node": "^4.0.0"
+  },
+  "scripts": {
+    "start": "node server.js",
+    "install": "node installAdapters.js",
+    "demo": "node demo/demoServer.js"
```
Comment (rcjsuen) [inline modules/orionode.debug.server/package.json:26]: How do I run/use this demo? I invoked `npm run demo` but wasn't sure what to do afterwards.

File: bundles/org.eclipse.orion.client.debug/build.properties [hunk 10]
```diff
@@ -0,0 +1,16 @@
+###############################################################################
+# Copyright (c) 2011 IBM Corporation and others.
```
Comment (rcjsuen) [inline bundles/org.eclipse.orion.client.debug/build.properties]: Please use 2017 here.

File: bundles/org.eclipse.orion.client.debug/pom.xml [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+    All rights reserved. This program and the accompanying materials
+    are made available under the terms of the Eclipse Public License v1.0
+    (http://www.eclipse.org/legal/epl-v10.html),
+    and the Eclipse Distribution License v1.0
+    (http://www.eclipse.org/org/documents/edl-v10.html).
+
+    Contributors:
+      Matthias Sohn - initial implementation
+-->
```
Comment (rcjsuen) [inline bundles/org.eclipse.orion.client.debug/pom.xml:13]: Please use 2017 here. I'm not sure how the copyright should be written regarding IBM/SAP for the header and the contributors section and such. I'm assuming you just copy/pasted this from another `pom.xml` file and made some modifications?

File: modules/orionode.debug.server/lib/protocol.js [hunk 12]
```diff
@@ -0,0 +1,82 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License.
+ *--------------------------------------------------------------------------------------------*/
```
Comment (rcjsuen) [inline modules/orionode.debug.server/lib/protocol.js:4]: What's the background story behind this? Sounds like a CQ may be needed here.

**Comment by rcjsuen** [comment]
You can't rename a launch configuration. If you try to edit it and then give it a new name, a new launch configuration will be saved with the new name instead of the existing one getting renamed.

**Comment by snorthov** [comment]
@rcjsuen I created a plan item for the work.  You can ask questions here or in https://bugs.eclipse.org/bugs/show_bug.cgi?id=515337

**Comment by carmacleod** [comment]
@rcjsuen This might be helpful: https://wiki.eclipse.org/Orion/Documentation/Developer_Guide/Debugger

**Comment by rcjsuen** [comment]
Thanks you two! I'll take another gander later...

**Comment by rcjsuen** [comment]
I think I'll keep my findings here and post any overarching comments in Bugzilla.

So I tried to follow the [article](https://wiki.eclipse.org/Orion/Documentation/Developer_Guide/Debugger) that Carolyn linked to but after clicking the run button it just did nothing before finally giving me a red error status message that said "timeout". Nothing was found in my Node logs or my browser logs. I'm running Node v6.9.5 on Windows 10.

Here is my folder structure:

```
git/orion.client/modules/orionode/server.js
git/orion.client/debug/ (root of my workspace)
git/orion.client/debug/tmp (newly created folder)
git/orion.client/debug/tmp/server.js (sample server.js per the steps in the article)
```

I ran the Node server like so...
```
> node modules/orionode/server.js -w debug
[2017-04-18 19:38:06.334] [INFO] [server] Using workspace: [...]
python cannot be loaded.
vscode-lldb cannot be loaded.
java cannot be loaded.
Loaded debug adapters: node.

[2017-04-18 19:38:39.102] [INFO] [server] Listening on port 8081...
```

File: bundles/org.eclipse.orion.client.debug/pom.xml (hunk 11 above)
Comment (MartinWangIBM) [reply bundles/org.eclipse.orion.client.debug/pom.xml:13]: Yes. Looks like every `pom.xml` uses the same copyright header.

File: modules/orionode.debug.server/lib/protocol.js (hunk 12 above)
Comment (MartinWangIBM) [reply modules/orionode.debug.server/lib/protocol.js:4]: This is the wire protocol behind the debugger. It's very similar to the VS Code language server one.
VS Code provides a node package with this wire protocol implementation (`vscode-debugadapter/lib/protocol` see below) to help debugger adapter developers. We use this here to implement the debugger host (the role of VS Code).
It contains three types of message: requests, responses and events. Requests and responses can be sent in both directions, however events can be only sent from adapters to the debugger host. As a result, this wire protocol implementation (`vscode-debugadapter/lib/protocol`) doesn't support receiving events, because an adapter doesn't need to.
Therefore we need to modify the source code of this implementation to support receiving events. The change is very small (line 52 - 54) but is nested deeply.
CQ has been approved.

**Comment by MartinWangIBM** [comment]
@rcjsuen 
> You can't rename a launch configuration. If you try to edit it and then give it a new name, a new launch configuration will be saved with the new name instead of the existing one getting renamed.

It is a bug of the run bar. I've created a new bug for it: https://bugs.eclipse.org/bugs/show_bug.cgi?id=515389

File: bundles/org.eclipse.orion.client.debug/web/README.md (hunk 8 above)
Comment (MartinWangIBM) [reply bundles/org.eclipse.orion.client.debug/web/README.md:62]: I think it's better to put all instructions in one place. Do you think it's a good idea to just write "See orion.client/bundles/org.eclipse.client.debug/web/README.md for installation instructions" in `orionode.debug.server`?

File: modules/orionode.debug.server/package.json (hunk 9 above)
Comment (MartinWangIBM) [reply modules/orionode.debug.server/package.json:26]: It will run on port 3000. It's a demo for the VS Code debug protocol. Basically you can write and send any debug request in JSON by your self, just like using a CLI debugger.

**Comment by MartinWangIBM** [comment]
@rcjsuen 
> So I tried to follow the article that Carolyn linked to but after clicking the run button it just did nothing before finally giving me a red error status message that said "timeout". Nothing was found in my Node logs or my browser logs. I'm running Node v6.9.5 on Windows 10.

Typically you will see the error message in the stderr of the node server. I guess the compilation fails. Check whether `orion.debug.server\adapters\node-debug\out` exists. If not, try to run `.\node_modules\.bin\tsc` in `orion.debug.server\adapters\node-debug`.

My setup is Windows 10, node 6.9.4, powershell with git which runs perfectly.

**Comment by rcjsuen** [comment]
Hi, @MartinWangIBM.

> Check whether `orion.debug.server\adapters\node-debug\out` exists.

It exists and I have a `node` and a `tests` folder inside of it.

```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       2017-04-16     06:27           2997 debugInjection.js
-a----       2017-04-16     06:27          18116 extension.js
-a----       2017-04-16     06:27          15257 extension.js.map
-a----       2017-04-16     06:27           1221 extension.nls.de.json
-a----       2017-04-16     06:27           1295 extension.nls.es.json
-a----       2017-04-16     06:27           1336 extension.nls.fr.json
-a----       2017-04-16     06:27           1281 extension.nls.it.json
-a----       2017-04-16     06:27           1632 extension.nls.ja.json
-a----       2017-04-16     06:27           1943 extension.nls.json
-a----       2017-04-16     06:27           1417 extension.nls.ko.json
-a----       2017-04-16     06:27           1976 extension.nls.ru.json
-a----       2017-04-16     06:27           1135 extension.nls.zh-cn.json
-a----       2017-04-16     06:27           1278 extension.nls.zh-tw.json
-a----       2017-04-16     07:45         152624 nodeDebug.js
-a----       2017-04-16     06:27         121810 nodeDebug.js.map
-a----       2017-04-16     06:27           3217 nodeDebug.nls.de.json
-a----       2017-04-16     06:27           3126 nodeDebug.nls.es.json
-a----       2017-04-16     06:27           3004 nodeDebug.nls.fr.json
-a----       2017-04-16     06:27           3326 nodeDebug.nls.it.json
-a----       2017-04-16     06:27           3847 nodeDebug.nls.ja.json
-a----       2017-04-16     06:27           4636 nodeDebug.nls.json
-a----       2017-04-16     06:27           3284 nodeDebug.nls.ko.json
-a----       2017-04-16     06:27           4847 nodeDebug.nls.ru.json
-a----       2017-04-16     06:27           2323 nodeDebug.nls.zh-cn.json
-a----       2017-04-16     06:27           2438 nodeDebug.nls.zh-tw.json
-a----       2017-04-16     06:27          12468 nodeV8Protocol.js
-a----       2017-04-16     06:27           9979 nodeV8Protocol.js.map
-a----       2017-04-16     06:27            110 nodeV8Protocol.nls.de.json
-a----       2017-04-16     06:27            143 nodeV8Protocol.nls.es.json
-a----       2017-04-16     06:27            133 nodeV8Protocol.nls.fr.json
-a----       2017-04-16     06:27             97 nodeV8Protocol.nls.it.json
-a----       2017-04-16     06:27            161 nodeV8Protocol.nls.ja.json
-a----       2017-04-16     06:27            203 nodeV8Protocol.nls.json
-a----       2017-04-16     06:27            122 nodeV8Protocol.nls.ko.json
-a----       2017-04-16     06:27            177 nodeV8Protocol.nls.ru.json
-a----       2017-04-16     06:27             99 nodeV8Protocol.nls.zh-cn.json
-a----       2017-04-16     06:27             99 nodeV8Protocol.nls.zh-tw.json
-a----       2017-04-16     06:27          10424 pathUtilities.js
-a----       2017-04-16     06:27           9252 pathUtilities.js.map
-a----       2017-04-16     06:27          21423 sourceMaps.js
-a----       2017-04-16     06:27          16823 sourceMaps.js.map
-a----       2017-04-16     06:27            187 terminateProcess.sh
-a----       2017-04-16     06:27           3946 URI.js
-a----       2017-04-16     06:27           3790 URI.js.map
```

```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----       2017-04-16     06:27          17765 adapter.test.js
-a----       2017-04-16     06:27          13897 adapter.test.js.map
-a----       2017-04-16     06:27          10835 pathUtilities.test.js
-a----       2017-04-16     06:27           9956 pathUtilities.test.js.map
-a----       2017-04-16     06:27           2837 URI.test.js
-a----       2017-04-16     06:27           2151 URI.test.js.map
```

**Comment by MartinWangIBM** [comment]
@rcjsuen 
Here are few more thing you can try:
* Try run `node nodeDebug.js` in `orion.debug.server\adapters\node-debug\out\node` to see if there is any error. It should have no output.
* Try run `npm test` in `orion.debug.server\adapters\node-debug`.

Unfortunately this "timeout" error is as annoying as segfault, because it doesn't provide any useful information. However it always happens when the adapter is not setup correctly. It happened several times to me when I tried to install new adapters. I'm going to write a "troubleshoot" page to document ways to debug.

File: bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js (hunk 2 above)
Comment (squarti) [reply bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js]: Could we create a drop down menu instead? It always shows "+", but if there are more deploy services, the user can choose which one (similar the the Push drop down in the git page)

File: bundles/org.eclipse.orion.client.ui/web/plugins/pageLinksPlugin.js (hunk 6 above)
Comment (squarti) [reply bundles/org.eclipse.orion.client.ui/web/plugins/pageLinksPlugin.js]: separate patch. We need to fix the electron build, because the shell requires native libraries. Similar to nodegit.

File: bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js (hunk 2 above)
Comment (MartinWangIBM) [reply bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js]: It will be a "dropdown inside a dropdown", which I don't think is good.

Another solution can be: If there is more than one deploy services and the user clicks "+", multiple buttons (like the current ones in this PR) shows and replace the "+" button.

Example:

| Runbar |
| --- |
| Config 1 |
| Config 2 |
| [+] |

After the user clicks the [+] button:

| Runbar |
| --- |
| Config 1 |
| Config 2 |
| [Service 1] [Service 2] |

File: bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js [hunk 13]
```diff
@@ -40,6 +41,7 @@ define([
 	 * @param options.preferencesService
 	 * @param options.statusService
 	 * @param options.actionScopeId
+	 * @param ootions.generalPreferences
```
Comment (bogg) [inline bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js]: options?

**Comment by rcjsuen** [comment]
Thanks, @MartinWangIBM. I ran `node nodeDebug.js` and nothing happened so I'm assuming that it's okay.

I ran `npm test` and got two failures related to launching (unsurprisingly).

```
  Node Debug Adapter
    basic
      √ unknown request should produce error (461ms)
    initialize
      √ should return supported features (481ms)
      √ should produce error for invalid 'pathFormat' (672ms)
    launch
      1) should run program to the end
      2) "after each" hook for "should run program to the end"

  pathUtilities
    normalize(path)
      √ should return a path with forward slashes and '..' removed
    join(absPath, relPath)
      √ should return a path with forward slashes
    isAbsolutePath(path)
      √ should return true when the path is absolute
      √ should return false when the path is relative
    makeRelative(target, path)
      √ identical paths should return empty string
      √ target and path same length
      √ target is longer
      √ path is longer
    makeRelative2(from, to)
      √ identical paths should return empty string
      √ from and to same length
      √ from is longer
      √ to is longer
    realPath(path)
      √ on a case insensitive file system realPath might return different casing for a given path
    normalizeDriveLetter
      √ should return an unmodified path if there is no drive letter
      √ should return a path with lower case drive letter
    multiGlob
      √ one pattern
      √ one pattern up
      √ one pattern with curly brackets
      √ one pattern with curly brackets, one exclude
      √ two patterns
      √ two include, one exclude
    multiGlobMatches
      √ simple
      √ braces
      √ negate braces
      √ include/exclude
    extendObject
      √ extend
      √ override
      √ extend by same
      √ extend by undefined

  URI
    file
      √ absolute unix paths
      √ absolute windows paths
      √ relative unix paths
      √ relative windows paths
      √ filePath


  37 passing (17s)
  2 failing

  1) Node Debug Adapter launch should run program to the end:
     Error: no event 'terminated' received after 5000 ms
      at Timeout._onTimeout (node_modules\vscode-debugadapter-testsupport\lib\debugClient.js:195:28)

  2) Node Debug Adapter "after each" hook for "should run program to the end":
     Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

File: bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js (hunk 13 above)
Comment (MartinWangIBM) [reply bundles/org.eclipse.orion.client.ui/web/orion/widgets/projects/RunBar.js]: oops..

**Comment by MartinWangIBM** [comment]
@rcjsuen 
This is really weird... Are you using node runtime managers such as `nvm` or `nvs`? Make sure `node` is exposed in `PATH`. If you suspect this being the problem, try to add a `runtimeExecutable` field in your launch configuration:

```json
{
    "program": "/YOUR/PROGRAM",
    "cwd": "/YOUR/CWD",
    "runtimeExecutable": "/PATH/TO/node"
}
```

**Comment by rcjsuen** [comment]
No, I don't use that stuff, @MartinWangIBM. I can run `npm` and `node` just fine from a regular ol' cmd command prompt. I tried your `runtimeExecutable` suggestion and it didn't help. Maybe I'll look at this some more another time...

File: bundles/org.eclipse.orion.client.javascript/web/javascript/hover.js [hunk 14]
```diff
@@ -436,8 +436,99 @@ Objects.mixin(JavaScriptHover.prototype, /** @lends javascript.JavaScriptHover.p
...
+/**
+ * @name javascript.JavaScriptDebugHoverEvaluationProvider
+ * @constructor
+ * @public
+ * @param {javascript.ASTManager} astManager
+ * @param {javascript.ScriptResolver} resolver
+ * @param {javascript.TernWorkerCore} ternWorker
+ * @param {javascript.CUProvider} cuProvider
+ */
+function JavaScriptDebugHoverEvaluationProvider(astManager, resolver, ternWorker, cuProvider) {
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.javascript/web/javascript/hover.js]: This name is silly. Any good idea?

File: bundles/org.eclipse.orion.client.ui/web/edit/setup.js [hunk 15]
```diff
@@ -1212,6 +1203,99 @@ objects.mixin(EditorViewer.prototype, {
 					}
 				}.bind(this));
 			}.bind(this));
+			// Listen annotation changes
```
Comment (MartinWangIBM) [inline bundles/org.eclipse.orion.client.ui/web/edit/setup.js]: @squarti PTAL this block of changes which handles annotation line changes and avoid duplicate events from split editor.

File: bundles/org.eclipse.orion.client.javascript/web/javascript/hover.js (hunk 14 above)
Comment (mrennie) [reply bundles/org.eclipse.orion.client.javascript/web/javascript/hover.js]: JavaScriptDebugHover would be better, but thats fine for now, we can change it later if we really don't like it

**Comment by mrennie** [comment]
+1 for the JavaScript plugin changes. It would be nice to just reuse the existing class and extend it to handle debug hovers, but having another class doesn't affect the functionality or add another dependency to JavaScriptPlugin.

**Comment by soulbeing** [comment]
PR move to https://github.com/eclipse/orion.client/pull/152, rebased and cleaned
